### PR TITLE
internal/blobstore: implement RemoveUpload

### PR DESCRIPTION
**Pull request by rogpeppe**
We have to be careful that we (probably RemoveExpiredUploads)
can't remove the parts for an upload just as someone has added the
upload to the store. Unfortunately this case is hard to test, because
it needs the FinishUpload to happen between RemoveUpload retrieving
the document and trying to remove it.

### CODE REVIEW

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/charmstore/2039/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/charmstore/2040/

File: internal/blobstore/multipart.go [hunk 1]
```diff
@@ -313,30 +314,100 @@ func (s *Store) setUploadHash(uploadId string, udoc *uploadDoc) (string, error)
...
+	if err != nil {
+		if errgo.Cause(err) == ErrNotFound {
+			return nil
+		}
+		return errgo.Mask(err)
+	}
+	return s.deleteUpload(udoc)
+}
+
+func (s *Store) deleteUpload(udoc *uploadDoc) error {
```
Comment (mhilton) [inline internal/blobstore/multipart.go]: make this removeUpload to be consistent?

File: internal/blobstore/multipart.go [hunk 2]
```diff
@@ -313,30 +314,100 @@ func (s *Store) setUploadHash(uploadId string, udoc *uploadDoc) (string, error)
...
+			removedDoc = true
+		case mgo.ErrNotFound:
+			// Someone called FinishUpload concurrently.
+			removeParts = false
+		default:
+			return errgo.Mask(err)
+		}
+	}
+	if !removedDoc {
+		if err := s.uploadc.RemoveId(udoc.Id); err != nil && err == mgo.ErrNotFound {
```
Comment (mhilton) [inline internal/blobstore/multipart.go]: err == mgo.ErrNotFound implies that it isn't nil doesn't it? Although I have a suspicion that's not the operator you wanted.

**Review by mhilton — APPROVED** [review]
LGMT with a couple of thoughts

File: internal/blobstore/multipart.go (hunk 2 above)
Comment (rogpeppe) [reply internal/blobstore/multipart.go]: Yes, should be err != mgo.ErrNotFound.

File: internal/blobstore/multipart.go (hunk 1 above)
Comment (rogpeppe) [reply internal/blobstore/multipart.go]: done

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/charmstore/2041/

**Review by fabricematrat — APPROVED** [review]

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/charmstore/2042/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/charmstore/2043/

**Comment by rogpeppe** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci-gce.jujugui.org:8080/job/charmstore-merge

**Pull request merged**